### PR TITLE
sync getLayerBlockIds concurrency 299

### PR DESCRIPTION
**Pull request by almogdepaz**

### CODE REVIEW

File: sync/sync.go [hunk 1]
```diff
@@ -135,8 +138,15 @@ func (s *Syncer) maxSyncLayer() uint32 {
 
 func (s *Syncer) Synchronise() {
 	for i := s.layers.LocalLayerCount(); i < s.maxSyncLayer(); {
+
 		i++
-		blockIds := s.getLayerBlockIDs(i) //returns a set of all known blocks in the mesh
+
+		blockIds, err := s.getLayerBlockIDs(i) //returns a set of all known blocks in the mesh
+		if err != nil {
+			log.Error("could not get layer block ids ", err)
```
Comment (almogdepaz) [inline sync/sync.go]: better error...

File: sync/sync.go [hunk 2]
```diff
@@ -246,33 +214,121 @@ func (s *Syncer) sendBlockRequest(peer Peer, id uint32) (chan Block, error) {
...
+	idSet := make(map[uint32]bool, s.config.layerSize)
+	//todo move this to config
+	ch := make(chan []uint32)
+	reqCounter := 0
+
+	for _, v := range m {
+		_, err := s.sendLayerIDsRequest(v, index, ch)
+		if err != nil {
+			log.Error("could not fetch layer ", index, " block ids from peer ", v) // todo recover from this
+			return nil, err
```
Comment (antonlerner) [inline sync/sync.go:238]: what happens to channel ch if the function is returned? no one reads from it...
will block other async requests

File: sync/sync.go [hunk 3]
```diff
@@ -246,33 +214,121 @@ func (s *Syncer) sendBlockRequest(peer Peer, id uint32) (chan Block, error) {
...
+	res := make(chan uint32, s.config.layerSize)
+	defer close(res)
+
+	for {
+		select {
+		case b := <-ch:
+			for _, id := range b {
+				if _, exists := idSet[id]; !exists {
+					idSet[id] = true
+					res <- id
```
Comment (antonlerner) [inline sync/sync.go]: if all b's > layerSize this will get struck trying to write to the cahnnel

File: sync/sync.go [hunk 4]
```diff
@@ -246,33 +214,121 @@ func (s *Syncer) sendBlockRequest(peer Peer, id uint32) (chan Block, error) {
...
+			}
+			reqCounter--
+			if reqCounter == 0 {
+				return res, nil
+			}
+		case <-timeout: // Got a timeout! fail with a timeout error
+			if len(res) == 0 {
+				return nil, errors.New("could not get block ids from any peer")
+			}
+			return res, nil
```
Comment (antonlerner) [inline sync/sync.go]: should this go here? what abour return statement at the end of the function? also, you've timed out here, shouldn't it be an error?

File: sync/sync.go (hunk 2 above)
Comment (almogdepaz) [reply sync/sync.go:238]: yes i know, theres a todo right next to it...
in the future we will find another peer with the same hash to poll from

File: sync/sync.go [hunk 5]
```diff
@@ -246,33 +214,121 @@ func (s *Syncer) sendBlockRequest(peer Peer, id uint32) (chan Block, error) {
...
+		}
+	}
+}
+
+func (s *Syncer) getLayerHashes(index uint32) (map[string]Peer, error) {
+	m := make(map[string]Peer, 20)
```
Comment (antonlerner) [inline sync/sync.go]: 20 is a very magical number... what does it mean?

Comment (almogdepaz) [reply sync/sync.go]: means this is waiting for integration with p2p enhancement to get connected peers

File: sync/sync.go [hunk 6]
```diff
@@ -246,33 +214,121 @@ func (s *Syncer) sendBlockRequest(peer Peer, id uint32) (chan Block, error) {
...
+		// Got a timeout! fail with a timeout error
+		case pair := <-ch:
+			m[string(pair.hash)] = pair.peer
+			resCounter++
+			if resCounter == totalRequests {
+				return m, nil
+			}
+		case <-timeout:
+			if len(m) > 0 {
+				return m, nil
```
Comment (antonlerner) [inline sync/sync.go]: is this a valid response?

File: sync/sync.go [hunk 7]
```diff
@@ -135,16 +137,22 @@ func (s *Syncer) maxSyncLayer() uint32 {
 
 func (s *Syncer) Synchronise() {
 	for i := s.layers.LocalLayerCount(); i < s.maxSyncLayer(); {
+
 		i++
-		blockIds := s.getLayerBlockIDs(i) //returns a set of all known blocks in the mesh
+
+		blockIds, err := s.getLayerBlockIDs(i) //returns a set of all known blocks in the mesh
+		if err != nil {
+			log.Error("could not get layer block ids: ", err)
```
Comment (antonlerner) [inline sync/sync.go:145]: use one of them for now

**Review by antonlerner — APPROVED** [review]

**Pull request merged**